src/Nethermind/Nethermind.BeaconNode/AttestationProducer.cs

```diff
@@ -0,0 +1,126 @@
+    public class AttestationProducer
+    {
+        private readonly BeaconChainUtility _beaconChainUtility;
+        private readonly BeaconStateAccessor _beaconStateAccessor;
+        private readonly BeaconStateTransition _beaconStateTransition;
+        private readonly ICryptographyService _cryptographyService;
+        private readonly IForkChoice _forkChoice;
+        private readonly ILogger _logger;
+        private readonly IStore _store;
+
+        public AttestationProducer(ILogger<AttestationProducer> logger,
+            ICryptographyService cryptographyService,
+            BeaconChainUtility beaconChainUtility,
+            BeaconStateAccessor beaconStateAccessor,
+            BeaconStateTransition beaconStateTransition,
+            IForkChoice forkChoice,
+            IStore store)
+        {
+            _logger = logger;
+            _cryptographyService = cryptographyService;
+            _beaconChainUtility = beaconChainUtility;
+            _beaconStateAccessor = beaconStateAccessor;
+            _beaconStateTransition = beaconStateTransition;
+            _forkChoice = forkChoice;
+            _store = store;
+        }
+
+        public async Task<Attestation> NewAttestationAsync(BlsPublicKey validatorPublicKey,
+            bool proofOfCustodyBit, Slot slot, CommitteeIndex index,
+            CancellationToken cancellationToken)
+        {
+            Root head = await _store.GetHeadAsync().ConfigureAwait(false);
+            BeaconBlock headBlock = (await _store.GetSignedBlockAsync(head).ConfigureAwait(false)).Message;
+            BeaconState parentState = await _store.GetBlockStateAsync(head).ConfigureAwait(false);
+
+            // Clone state (will mutate) and process outstanding slots
+            BeaconState headState = BeaconState.Clone(parentState);
+            _beaconStateTransition.ProcessSlots(headState, slot);
+
+            // Set attestation_data.index = index where index is the index associated with the validator's committee.
+            ValidatorIndex? validatorIndex = FindValidatorIndexByPublicKey(headState, validatorPublicKey);
+
+            if (validatorIndex == null)
+            {
+                throw new Exception($"Can not find validator index for public key {validatorPublicKey}");
+            }
+
+            // TODO: May need a more efficient way to try and find the committee and position within the committee.
+            // Some of this may already be cached in Validator Assignments (generally stable for an epoch),
+            // but not with the index within the committee. Easy enough to extend and use the same cache.
+
+            IReadOnlyList<ValidatorIndex> committee =
+                _beaconStateAccessor.GetBeaconCommittee(headState, headState.Slot, index);
+            int committeeSize = committee.Count;
+
+            int? committeeMemberIndexOfValidator = null;
+            for (int committeeMemberIndex = 0; committeeMemberIndex < committee.Count; committeeMemberIndex++)
```

Comment:
I think foreach is far better when you need to enumerate through whole collection
time is practically the same and the readability goes up